File: ProjectTemplateSourceSolutions/VisualBasic/RazorViewEngine/Containers/EndpointContainer.vb
Imports Junior.Route.ViewEngines.Razor.CompiledTemplateFactories
Imports Junior.Route.ViewEngines.Razor.CodeDomProviderFactories
Imports Junior.Route.AspNetIntegration
Imports Junior.Common
Imports JuniorRouteWebApplication.Endpoints
Imports Junior.Route.AutoRouting.Containers
Imports Junior.Route.Routing
Imports Junior.Route.ViewEngines.Razor.Routing.TemplatePathResolvers
Imports Junior.Route.ViewEngines.Razor.TemplateAssemblyReferenceResolvers
Imports Junior.Route.ViewEngines.Razor.TemplateCompilers
Imports Junior.Route.ViewEngines.Razor.TemplateClassNameBuilders
Imports Junior.Route.ViewEngines.Razor.TemplateCodeBuilders
Imports Junior.Route.ViewEngines.Razor.Routing.TemplateRepositories

Namespace Containers
	Public Class EndpointContainer
		Implements IContainer

		Private ReadOnly _fileSystemRepository As FileSystemRepository

		Public Sub New(httpRuntime As IHttpRuntime, guidFactory As IGuidFactory)
			Dim pathResolver = New VisualBasicResolver(httpRuntime)
			Dim fileSystem = New FileSystem(httpRuntime)
			Dim appDomainResolver = New AppDomainResolver
			Dim compiler = New TemplateCompiler(appDomainResolver)
			Dim classNameBuilder = New RandomGuidBuilder(guidFactory)
			Dim codeBuilder = New VisualBasicBuilder
			Dim codeDomProviderFactory = New FileExtensionFactory
			Dim compiledTemplateFactory = New ActivatorFactory

			_fileSystemRepository = New FileSystemRepository(pathResolver, fileSystem, compiler, classNameBuilder, codeBuilder, codeDomProviderFactory, compiledTemplateFactory)
		End Sub

		Public Function GetInstance(Of T)() As T Implements IContainer.GetInstance
			Return GetInstance(GetType(T))
		End Function

		Public Function GetInstance(ByVal type As Type) As Object Implements IContainer.GetInstance
			If (type = GetType(HelloWorld)) Then
				Return New HelloWorld(_fileSystemRepository)
			End If

			Throw New ArgumentException(String.Format("Could not get instance for {0}.", type.FullName), "type")
		End Function
	End Class
End Namespace

File: ProjectTemplateSourceSolutions/CSharp/RazorViewEngine/Containers/EndpointContainer.cs
using System;
using System.Configuration;
using System.Web.Configuration;

using Junior.Common;
using Junior.Route.AspNetIntegration;
using Junior.Route.AutoRouting.Containers;
using Junior.Route.Routing;
using Junior.Route.ViewEngines.Razor.CodeDomProviderFactories;
using Junior.Route.ViewEngines.Razor.CompiledTemplateFactories;
using Junior.Route.ViewEngines.Razor.Routing.TemplatePathResolvers;
using Junior.Route.ViewEngines.Razor.Routing.TemplateRepositories;
using Junior.Route.ViewEngines.Razor.TemplateAssemblyReferenceResolvers;
using Junior.Route.ViewEngines.Razor.TemplateClassNameBuilders;
using Junior.Route.ViewEngines.Razor.TemplateCodeBuilders;
using Junior.Route.ViewEngines.Razor.TemplateCompilers;

using JuniorRouteWebApplication.Endpoints;

namespace JuniorRouteWebApplication.Containers
{
	public class EndpointContainer : IContainer
	{
		private readonly FileSystemRepository _fileSystemRepository;

		public EndpointContainer(IHttpRuntime httpRuntime, IGuidFactory guidFactory)
		{
			var pathResolver = new CSharpResolver(httpRuntime);
			var fileSystem = new FileSystem(httpRuntime);
			var appDomainResolver = new AppDomainResolver();
			var compiler = new TemplateCompiler(appDomainResolver);
			var classNameBuilder = new RandomGuidBuilder(guidFactory);
			var codeBuilder = new CSharpBuilder();
			var codeDomProviderFactory = new FileExtensionFactory();
			var compiledTemplateFactory = new ActivatorFactory();
			var compilationSection = (CompilationSection)ConfigurationManager.GetSection("system.web/compilation");
			bool reloadChangedTemplateFiles = compilationSection != null && compilationSection.Debug;
			var fileSystemRepositoryConfiguration = new FileSystemRepositoryConfiguration(reloadChangedTemplateFiles);

			_fileSystemRepository = new FileSystemRepository(
				pathResolver,
				fileSystem,
				compiler,
				classNameBuilder,
				codeBuilder,
				codeDomProviderFactory,
				compiledTemplateFactory,
				fileSystemRepositoryConfiguration);
		}

		public T GetInstance<T>()
		{
			return (T)GetInstance(typeof(T));
		}

		public object GetInstance(Type type)
		{
			if (type == typeof(HelloWorld))
			{
				return new HelloWorld(_fileSystemRepository);
			}

			throw new ArgumentException(String.Format("Could not get instance for {0}.", type.FullName), "type");
		}
	}
}